src/Illuminate/Foundation/Testing/TestCase.php

```diff
@@ -262,4 +269,116 @@ protected function callBeforeApplicationDestroyedCallbacks()
             }
         }
     }
+
+    /**
+     * This method is called when a test method did not execute successfully.
+     *
+     * @param  \Throwable  $exception
+     * @return void
+     */
+    protected function onNotSuccessfulTest(Throwable $exception): void
+    {
+        if (! $exception instanceof ExpectationFailedException || is_null($this->latestResponse)) {
+            parent::onNotSuccessfulTest($exception);
+        }
+
+        if ($lastException = $this->latestResponse->exceptions->last()) {
+            parent::onNotSuccessfulTest($this->appendExceptionToException($lastException, $exception));
+
+            return;
+        }
+
+        if ($this->latestResponse->baseResponse instanceof RedirectResponse) {
+            $session = $this->latestResponse->baseResponse->getSession();
+
+            if (! is_null($session) && $session->has('errors')) {
+                parent::onNotSuccessfulTest($this->appendErrorsToException($session->get('errors')->all(), $exception));
+
+                return;
+            }
+        }
+
+        if ($this->latestResponse->baseResponse->headers->get('Content-Type') === 'application/json') {
+            $testJson = new AssertableJsonString($this->latestResponse->getContent());
+
+            if (isset($testJson['errors'])) {
+                parent::onNotSuccessfulTest($this->appendErrorsToException($testJson->json(), $exception, true));
+
+                return;
+            }
+        }
+
+        parent::onNotSuccessfulTest($exception);
+    }
+
+    /**
+     * Append an exception to the message of another exception.
+     *
+     * @param  \Throwable  $exceptionToAppend
+     * @param  \Throwable  $exception
+     * @return \Throwable
+     */
+    protected function appendExceptionToException($exceptionToAppend, $exception)
+    {
+        $exceptionMessage = $exceptionToAppend->getMessage();
+        $exceptionToAppend = (string) $exceptionToAppend;
+        $message = <<<"EOF"
+            The following exception occurred during the last request:
+
+            $exceptionToAppend
+
+            ----------------------------------------------------------------------------------
+
+            $exceptionMessage
+            EOF;
+
+        return $this->appendMessageToException($message, $exception);
+    }
+
+    /**
+     * Append errors to an exception message.
+     *
+     * @param  array  $errors
+     * @param  \Throwable  $exception
+     * @param  bool  $json
+     * @return \Throwable
+     */
+    protected function appendErrorsToException($errors, $exception, $json = false)
+    {
+        $errors = $json
+            ? json_encode($errors, JSON_PRETTY_PRINT)
+            : implode(PHP_EOL, Arr::flatten($errors));
+
+        // JSON error messages may already contain the errors, so we shouldn't duplicate them.
+        if (str_contains($exception->getMessage(), $errors)) {
```

Comment:
Does this works? Asking because it seems we change the format of the JSON string by using the `JSON_PRETTY_PRINT` flag.

Reply:
Yeah - the existing output in the exception is also pretty printed